content/en/blog/_posts/2021-05-04-writing-a-controller-for-pod-labels.md

```diff
@@ -0,0 +1,469 @@
+    if err := r.Get(ctx, req.NamespacedName, &pod); err != nil {
+        log.Error(err, "unable to fetch Pod")
+        return ctrl.Result{}, err
+    }
+
+    return ctrl.Result{}, nil
+}
+```
+
+Our `Reconcile` method will be called when a Pod is created, updated, or
+deleted. In the deletion case, our call to `r.Get` will return a specific error.
+Let's import the package that defines this error:
+
+```go
+package controllers
+
+import (
+    // other imports...
+    apierrors "k8s.io/apimachinery/pkg/api/errors"
+    // other imports...
+)
+```
+
+We can now handle this specific error and — since our controller does not care
+about deleted pods — explicitly ignore it:
+
+```go
+    /*
+        Step 0: Fetch the Pod from the Kubernetes API.
+    */
+
+    var pod corev1.Pod
+    if err := r.Get(ctx, req.NamespacedName, &pod); err != nil {
+        if apierrors.IsNotFound(err) {
+            // we'll ignore not-found errors, since we can get them on deleted requests.
+            return ctrl.Result{}, nil
+        }
+        log.Error(err, "unable to fetch Pod")
+        return ctrl.Result{}, err
+    }
+```
+
+Next, lets edit our Pod so that our dynamic label is present if and only if our
+annotation is present:
+
+```go
+    /*
+        Step 1: Add or remove the label.
+    */
+
+    labelShouldBePresent := pod.Annotations[addPodNameLabelAnnotation] == "true"
+    labelIsPresent := pod.Labels[podNameLabel] == pod.Name
+
+    if labelShouldBePresent == labelIsPresent {
+        // The desired state and actual state of the Pod are the same.
+        // No further action is required by the operator at this moment.
+        log.Info("no update required")
+        return ctrl.Result{}, nil
+    }
+
+    if labelShouldBePresent {
+        // If the label should be set but is not, set it.
+        if pod.Labels == nil {
+            pod.Labels = make(map[string]string)
+        }
+        pod.Labels[podNameLabel] = pod.Name
+        log.Info("adding label")
+    } else {
+        // If the label should not be set but is, remove it.
+        delete(pod.Labels, podNameLabel)
+        log.Info("removing label")
+    }
+```
+
+Finally, let's push our updated Pod to the Kubernetes API:
+
+```go
+    /*
+        Step 2: Update the Pod in the Kubernetes API.
+    */
+
+    if err := r.Update(ctx, &pod); err != nil {
+        log.Error(err, "unable to update Pod")
+        return ctrl.Result{}, err
+    }
+```
+
+When writing our updated Pod to the Kubernetes API, there is a risk that the Pod
+has been updated or deleted since we first read it. When writing a Kubernetes
+controller, we should keep in mind that we are not the only actors in the
+cluster. When this happens, the best thing to do is start the reconciliation
+from scratch, by requeuing the event. Lets do exactly that:
+
+```go
+    /*
+        Step 2: Update the Pod in the Kubernetes API.
+    */
+
+    if err := r.Update(ctx, &pod); err != nil {
+        if apierrors.IsConflict(err) {
```

Comment:
Consider demonstrating the new SSA apply client? @jpbetz is there a good doc to look at for that? It would make this update code simpler.

Reply:
I need to add some better documentation in client-go (I've created a task for myself).

For this case, I think that the apply equivalent for Step 1 should be roughly:

```go
import (
         ...
         v1ac "k8s.io/client-go/applyconfigurations/core/v1"
)

...

var podApplyConfig *v1ac.PodApplyConfiguration
if if labelShouldBePresent {
     podApplyConfig = v1ac.Pod(pod.Name, pod.Namespace).
         WithLabels(map[string]string{"podNameLabel", pod.Name})
} else {
     podApplyConfig = v1ac.Pod(pod.Name, pod.Namespace).
         WithLabels(map[string]string{})
}

...

r.Apply(ctx, podApplyConfig, metav1.ApplyOptions{FieldManager: "<TODO:put field manager name here>", Force: true})
```

Reply:
I love the idea of using this apply client!

The Operator SDK makes each controller implement the `sigs.k8s.io/controller-runtime/pkg/client.Client` interface which does not have an `Apply` method.

I think that configuring the controller to implement this method would be too complex for this introductory blog post.